src/Microsoft.Diagnostics.Monitoring.WebApi/CollectionRulePipelineState.cs

```diff
@@ -0,0 +1,238 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
+
+namespace Microsoft.Diagnostics.Monitoring.WebApi
+{
+    internal class CollectionRulePipelineState
+    {
+        public CollectionRuleState CurrentState { get; private set; } = CollectionRuleState.Running;
+        public string CurrentStateReason { get; private set; } = Strings.Message_CollectionRuleStateReason_Running;
+        public Queue<DateTime> ExecutionTimestamps { get; set; }
+        public List<DateTime> AllExecutionTimestamps { get; set; }
+        public TimeSpan? ActionCountSlidingWindowDuration { get; private set; }
+        public TimeSpan? RuleDuration { get; private set; }
+        public int ActionCountLimit { get; private set; }
+
+        private readonly object _lock = new object();
```

Comment:
I think this lock object should be in CollectionRulePipeline: the pipeline locks whenever it needs to mutate the state object and when it is making a copy of the state object to hand out to external callers. Then the object itself doesn't need to have any synchronization since we don't have multiple threads or callers accessing the same instance.

Reply:
Discussed in person: add comments as to why its preferable to have it here